src/asm/fstack.c

```diff
@@ -420,8 +430,53 @@ void fstk_RunRept(uint32_t count, int32_t reptLineNo, char *body, size_t size)
 		fatalerror("Failed to set up lexer for rept block\n");
 	lexer_SetStateAtEOL(contextStack->lexerState);
 	contextStack->uniqueID = macro_UseNewUniqueID();
+	return true;
+}
+
+void fstk_RunRept(uint32_t count, int32_t reptLineNo, char *body, size_t size)
+{
+	dbgPrint("Running REPT(%" PRIu32 ")\n", count);
+
+	if (count == 0)
+		return;
+	if (!newReptContext(reptLineNo, body, size))
+		return;
+
 	contextStack->nbReptIters = count;
+	contextStack->foreachName = NULL;
+}
+
+void fstk_RunForeach(char const *symName, int32_t start, int32_t stop, int32_t step,
+		     int32_t reptLineNo, char *body, size_t size)
+{
+	dbgPrint("Running FOREACH(\"%s\", %" PRId32 ", %" PRId32 ", %" PRId32 ")\n",
+		 symName, start, stop, step);
+
+	struct Symbol *sym = sym_AddSet(symName, start);
+
+	if (sym->type != SYM_SET)
+		return;
+
+	uint32_t count = 0;
+
+	if (step > 0 && start < stop)
+		count = (stop - start - 1) / step + 1;
+	else if (step < 0 && stop < start)
+		count = (start - stop - 1) / -step + 1;
+	else
+		error("FOREACH cannot have a step value of 0\n");
```

Comment:
Wouldn't this also trigger for `1, 2, -1` with the same message? (This may warrant additional test cases, too)

Reply:
You're right, the message should only show for `step == 0`. Other cases like `1, 2, -1` or `5, 5, 1` will still have `count == 0` and will just do nothing, like `rept 0`.

Reply:
Added test cases for this, and fixed the macOS newline issue.